rpc/core/env.go

```diff
@@ -100,6 +101,65 @@ type Environment struct {
 	genChunks []string
 }
 
+// InitGenesisChunks checks whether it makes sense to create a cache of chunked
+// genesis data. It is called once on Node startup.
+// Rules of chunking:
+// - if the genesis file's size is <= 16MB, then no chunking. An `Environment`
+// object will store a pointer to the genesis in its GenDoc field. Its genChunks
+// field will be set to nil. `/genesis` RPC API will return the GenesisDoc itself.
+// - if the genesis file's size is > 16MB, then use chunking. An `Environment`
+// object will store a slice of base64-encoded chunks in its genChunks field. Its
+// GenDoc field will be set to nil. `/genesis` RPC API will redirect users to use
+// the `/genesis_chunked` API.
+func (env *Environment) InitGenesisChunks() error {
+	if len(env.genChunks) > 0 {
+		// we already computed the chunks, return.
+		return nil
+	}
+
+	if env.GenDoc == nil {
+		// chunks not computed yet, but no genesis available.
+		// This should not happen.
+		return errors.New("pointer to genesis is nil")
+	}
+
+	data, err := cmtjson.Marshal(env.GenDoc)
+	if err != nil {
+		return err
+	}
+
+	// If genesis is less than 16MB, then no chunking.
+	// Keep a pointer to a GenesisDoc in env.GenDoc.
+	if len(data) <= genesisChunkSize {
+		env.genChunks = nil
```

Comment:
Does that mean we will be hitting this code path again and again because `if len(env.genChunks) > 0 {` only checks that N of chunks is > 0, but not when we don't need them?

Reply:
I'm not sure I understand the question 🤔 
- We don't create chunks if the genesis size is less than 16 MB. Therefore, to understand whether we need to chunk or not, we have to compute the size of the genesis.
- The implementation that we have in production computes the size of the genesis from the `[]byte` slice obtained by calling `json.Marshal`, rather than the `[]byte` slice obtained just after reading the file contents. To keep the existing behavior, the changes in this PR also compute the size from the serialized genesis.
- `InitGenesisChunks` is called only once from `Node.ConfigureRPC()` at startup, so the check `if len(env.genChunks) > 0` is performed only once.
- We initialize and modify `env.genChunks` only in `InitGenesisChunks`. Since this field is unexported, there's no risk of accidental initialization elsewhere.

We can discuss whether computing the genesis's size from the serialized `[]byte` blob rather than directly reading the file is the best approach. Serialization removes extra characters (e.g., spaces, tabs), thus reducing the input size. For example, in tests, I used a ~69 MB genesis file that reduced to ~34 MB when serialized. That's a big difference: using the file size would create 5 chunks, but using the serialized size results in 3 chunks. Since we use the JSON `[]byte` blob in production, I preferred not to change this behavior.

Does this make sense? Or is your question about something different?